Rework status codes

mikekistler opened the pull request:
This PR illustrates the change I am proposing in #197.  Some work is still needed to polish this up but I think the key features are there -- enough to illustrate the proposal.

Key elements:
- the http library now validates and processes the target property of the `@statusCode` decorator.  The property value must be a numeric or string literal or union of numeric or string literals which are three digit codes starting with 1-5.
- the openapi library implements a new decorator `@default` (might want a different name) to specify that the decorated model should be the "default" response of an operation.

I did not implement decorators for "2XX", "4XX", or "5XX" response codes, but these would be very similar to the `@default` decorator.

nguerrera commented:
> `@default` (might want a different name)

Maybe `@defaultResponse`?

nguerrera commented:
This looks good to me. I presume there will be fallout in the closed repo if we merge this, right? Would you be able to prep the reaction ahead of time so we can review that? Last time we moved around @statusCode's in standard lib, the flow to closed repo was blocked for a while. You should be able to make parallel changes following https://github.com/Azure/cadl-azure/blob/main/CONTRIBUTING.md#making-a-cross-cutting-change-across-both-repos

timotheeguerin commented on packages/openapi3/test/test-return-types.ts:
```diff
@@ -353,13 +328,9 @@ describe("openapi3: return types", () => {
     `
     );
     strictEqual(diagnostics.length, 3);
```
you can use the new `expectDiagnostics` util function to pass the list of diagnostic that it should expect

timotheeguerin commented on packages/openapi/src/decorators.ts:
```diff
@@ -29,3 +29,22 @@ export function $extension(program: Program, entity: Type, extensionName: string
+
+// The @default decorator can be applied to a model. When that model is used
+// as the return type of an operation, this return type will be the default response.
+const defaultResponses = new Set<Type>();
```
Should use the program to keep state not in global objects. 
```ts
program.stateSet(uniqueSymbolForThisDecorator)
```

timotheeguerin commented on packages/openapi/src/decorators.ts:
```diff
@@ -29,3 +29,22 @@ export function $extension(program: Program, entity: Type, extensionName: string
+const defaultResponses = new Set<Type>();
+export function $default(program: Program, entity: Type) {
+  if (entity.kind !== "Model") {
+    reportDiagnostic(program, {
```
Depending on if this get merged first or the decorator test/validation https://github.com/microsoft/cadl/pull/236 there is a `validateDecoratorTarget` util function that does that and emit the diagnostic for you. 
"decorator-wrong-type" has also been removed from library

timotheeguerin reviewed (commented):
Just a few nits

mikekistler commented:
I've addressed @timotheeguerin 's comments, changed the decorator name to `@defaultResponse`, and rebased onto the current main.

I did not change the `@error` decorator name since I think we are still deciding if that should move to the Azure library.

I'll try to have the changed ready in cadl-autorest by tomorrow at the latest.

mikekistler commented on packages/openapi3/src/openapi.ts:
```diff
@@ -450,95 +458,21 @@ function createOAPIEmitter(program: Program, options: OpenAPIEmitterOptions) {
-  // Note: these descriptions come from https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html
-  function getDescriptionForStatusCode(statusCode: string) {
-    switch (statusCode) {
-      case "2XX":
-        return "Successful";
-      case "200":
-        return "Ok";
-      case "201":
-        "Created";
-      case "202":
-        return "Accepted";
-      case "204":
-        return "No Content";
-      case "3XX":
-        return "Redirection";
-      case "301":
-        return "Moved Permanently";
-      case "304":
-        return "Not Modified";
-      case "4XX":
-        return "Client Error";
-      case "400":
-        return "Bad Request";
-      case "401":
-        return "Unauthorized";
-      case "403":
-        return "Forbidden";
-      case "404":
-        return "Not Found";
-      case "409":
-        return "Conflict";
-      case "412":
-        return "Precondition Failed";
-      case "5XX":
-        return "Server Error";
-      case "default":
-        return "An unexpected error response";
-    }
-    // We might want to throw here -- rather than giving some default
-    return "A successful response";
```
I moved the status code description method to the http library and improved the handling of codes not explicitly listed.

mikekistler commented:
@markcowl I addressed your comments.  Would you mind taking one more look for I merge?
